app/controllers/application_controller.rb

```diff
@@ -5,10 +5,31 @@ class ApplicationController < ActionController::Base
   # For APIs, you may want to use :null_session instead.
   protect_from_forgery with: :exception
 
+  protected
   # Use callbacks to share common setup or constraints between actions.
   def set_vault
-    @vault = params[:vault_id] ? Vault.find_by_id(params[:vault_id]) : Vault.find_by_id(params[:id])
-    redirect_to root_url if @vault.nil? || @vault.user_username != params[:username]
+    @vault = params[:vault_id] ? Vault.find_vault(params[:vault_id]) : Vault.find_vault(params[:id])
```

Comment:
just made an update. i remove the authenticate username and created a simple `if..else` if the action is show authenticate both @vault and @vault.user_username if not just authenticate @vault it is nil.